Modules/arraymodule.c

```diff
@@ -2833,11 +3231,15 @@ array_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 
                     if (n > 0) {
                         arrayobject *self = (arrayobject *)a;
-                        // self->ob_item may be NULL but it is safe.
-                        PyMem_Free(self->ob_item);
-                        self->ob_item = (char *)ustr;
+                        self->data = arraydata_set_items(self->data,
```

Comment:
This is a bit confusing and `self->data` is always NULL here. The `len > 0` cases are handled on 3915 and on 3267.

Instead:

```
assert(self->data == NULL);
self->data = arraydata_alloc(n, sizeof(wchar_t));
```

and get rid of the `arraydata_set_items` function.